File: app/utils/FileUtil.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2012 NAVER Corp.
 * http://yobi.io
 *
 * @author Ahn Hyeok Jun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package utils;

import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang.StringUtils;
import org.apache.tika.Tika;
import org.apache.tika.metadata.Metadata;
import org.apache.tika.mime.MediaType;
import org.mozilla.universalchardet.UniversalDetector;

import java.io.*;
import java.nio.charset.Charset;
import java.nio.file.Path;

public class FileUtil {

    public static void rm_rf(File file) throws Exception {
        if(file.isDirectory()){
            File[] list = file.listFiles();
            if (list == null) {
                throw new Exception("Unexpected error while deleting: " + file);
            }
            for (File f : list) {
                rm_rf(f);
            }
            file.delete();
        } else {
            file.delete();
        }
    }

    static private String or(String a, String b) {
        return StringUtils.isNotEmpty(a) ? a : b;
    }

    /**
     * Detect the charset used for the given byte array.
     *
     * Return "UTF-8" if it fails.
     *
     * We use mozilla's UniversalDetector instead of Tika's CharsetDetector
     * which raises "mark/reset not supported" IOException for FileInputStream.
     *
     * @param bytes - a byte array to be checked
     * @return charset - the charset used
     * @throws IOException
     */
    public static String detectCharset(byte bytes[]) {
        UniversalDetector detector = new UniversalDetector(null);
        int offset = 0;

        do {
            int blockSize = Math.min(4096, bytes.length - offset);
            detector.handleData(bytes, offset, blockSize);
            offset += blockSize;
        } while(offset < bytes.length);
        detector.dataEnd();

        return or(detector.getDetectedCharset(), "UTF-8");
    }

    /**
     * Detect the charset used for the given stream.
     *
     * Return "UTF-8" if it fails.
     *
     * We use mozilla's UniversalDetector instead of Tika's CharsetDetector
     * which raises "mark/reset not supported" IOException for FileInputStream.
     *
     * @param is - an input stream to be checked
     * @return charset - the charset used
     * @throws IOException
     */
    public static String detectCharset(InputStream is) throws IOException {
        UniversalDetector detector = new UniversalDetector(null);
        byte[] buf = new byte[4096];
        int nRead;

        while ((nRead = is.read(buf)) > 0 && !detector.isDone()) {
            detector.handleData(buf, 0, nRead);
        }
        detector.dataEnd();

        return or(detector.getDetectedCharset(), "UTF-8");
    }

    public static MediaType detectMediaType(File file, String name) throws IOException {
        return detectMediaType(new FileInputStream(file), name);
    }

    public static MediaType detectMediaType(byte[] bytes, String name) throws IOException {
        return detectMediaType(new ByteArrayInputStream(bytes), name);
    }

    /**
     * Detects media type of the given resource, by using Apache Tika.
     *
     * This method does following additional tasks besides Tika:
     * 1. Adds a charset parameter for text resources.
     * 2. Fixes Tika's misjudge of media type for ogg videos
     *
     * @param is    the input stream to read the resource
     * @param name  the filename of the resource
     * @return the detected media type which optionally includes a charset parameter
     *         e.g. "text/plain; charset=utf-8"
     * @throws IOException
     */
    public static MediaType detectMediaType(InputStream is, String name)
        throws IOException {
        Metadata meta = new Metadata();
        meta.add(Metadata.RESOURCE_NAME_KEY, name);
        MediaType mediaType = new Tika().getDetector().detect(
                new BufferedInputStream(is), meta);

        if (mediaType.getType().toLowerCase().equals("text")) {
            return new MediaType(mediaType, Charset.forName(FileUtil.detectCharset(is)));
        } else if (mediaType.equals(MediaType.audio("ogg"))
                && FilenameUtils.getExtension(name).toLowerCase().equals("ogv")) {
            // This fixes Tika's misjudge of media type for ogg videos.
            return new MediaType("video", "ogg");
        }

        return mediaType;
    }

    public static String getCharset(MediaType mediaType) {
        return mediaType.hasParameters()
                ? mediaType.getParameters().get("charset") : null;
    }

    /**
     * Checks whether the subpath is a subpath of the given path
     *
     * @param subpath
     * @param path
     * @return true if the subpath is a subpath of the given path
     * @throws IOException
     */
    public static boolean isSubpathOf(Path subpath, Path path) throws IOException {
        return isSubpathOf(subpath, path, true);
    }

    public static boolean isSubpathOf(Path subpath, Path path, boolean resolveSymlink) throws IOException {
        if (resolveSymlink) {
            path = path.toRealPath();
            subpath = subpath.toRealPath();
        }

        return subpath.normalize().startsWith(path.normalize());
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/support/FileUtil.kt
package com.github.yonaprojects.yona.domain.support

import org.apache.tika.Tika
import org.apache.tika.metadata.Metadata
import org.apache.tika.metadata.TikaCoreProperties
import org.mozilla.universalchardet.UniversalDetector
import java.io.BufferedInputStream
import java.io.File
import java.io.InputStream
import java.nio.charset.Charset

object FileUtil {

    fun detectCharset(bytes: ByteArray): String {
        val detector = UniversalDetector(null)
        var offset = 0

        do {
            val blockSize = Math.min(4096, bytes.size - offset)
            detector.handleData(bytes, offset, blockSize)
            offset += blockSize
        } while (offset < bytes.size)
        detector.dataEnd()

        return detector.detectedCharset ?: "UTF-8"
    }

    fun detectCharset(inputStream: InputStream): String {
        val detector = UniversalDetector(null)
        val buf = ByteArray(4096)
        var nRead: Int

        while (inputStream.read(buf).also { nRead = it } > 0 && !detector.isDone) {
            detector.handleData(buf, 0, nRead)
        }
        detector.dataEnd()

        return detector.detectedCharset ?: "UTF-8"
    }

    // yona FileUtil.detectMediaType() 대응. Apache Tika로 실제 파일 내용(매직
    // 바이트)을 콘텐츠 기반으로 감지한다 — 원본 파일명(name)은 힌트로만 쓰이고, 확장자가 없는
    // 해시 파일명이어도 정확히 감지된다(JDK Files.probeContentType()은 사실상 확장자 기반이라
    // 해시 파일명에서 거의 항상 감지에 실패한다).
    fun detectMediaType(file: File, name: String): String {
        val metadata = Metadata()
        metadata.set(TikaCoreProperties.RESOURCE_NAME_KEY, name)
        // Tika 4.x부터 저수준 Detector SPI(Tika().detector.detect())의 시그니처가 바뀌어서
        // (TikaInputStream + 추가 파라미터 요구) 내부 SPI에 직접 의존하지 않는 상위 Tika 파사드의
        // detect(InputStream, Metadata) 편의 메서드로 대체한다 — 반환 타입이 MediaType이 아니라
        // String(예: "text/plain")이라 아래 주타입 판별을 substringBefore로 바꿨다.
        val mediaType = file.inputStream().use { input ->
            Tika().detect(BufferedInputStream(input), metadata)
        }

        return when {
            mediaType.substringBefore('/').lowercase() == "text" -> {
                val charset = file.inputStream().use { detectCharset(it) }
                "$mediaType; charset=${Charset.forName(charset).name()}"
            }
            // Tika가 ogg 비디오를 audio/ogg로 오판하는 것을 보정 (yona FileUtil 동일 대응).
            mediaType == "audio/ogg" && name.substringAfterLast('.', "").lowercase() == "ogv" ->
                "video/ogg"
            else -> mediaType
        }
    }
}
